File: javascript/luhn/luhn.js
function doubleThis(array) {
  var output = [array[0]];
  for (var i = 1; i < array.length; i++) {
    var x = array[i];
    if (i % 2 !== 0) {
      x *= 2
      x = x > 9 ? x - 9 : x;
    }
    output.push(x);
  }
  return output.reverse();
};

function Luhn(num) {
  num = num.toString().split("").map(x => parseInt(x)).reverse();
  this.checkDigit = num[0];
  this.addends = doubleThis(num);
  this.checksum = this.addends.reduce((p, c) => p + c);
  this.valid = this.checksum % 10 === 0;
};

Luhn.create = function(num) {
  num *= 10;
  var luhn = new Luhn(num);
  return luhn.valid ? num : num + (10 - (luhn.checksum % 10));
};

module.exports = Luhn;

File: csharp/luhn/Luhn.cs
using System;
using System.Linq;

public class Luhn
{
  private int[] numbers;

  public Luhn(long number)
  {
    numbers = number.ToString().Select(x => (int)Char.GetNumericValue(x)).ToArray();
  }

  public int CheckDigit
  {
    get
    {
      return numbers.Last();
    }
  }

  public int[] Addends
  {
    get
    {
      return Doubled(numbers);
    }
    
  }

  public int Checksum
  {
    get
    {
      return Addends.Sum();
    }
  }

  public bool Valid
  {
    get
    {
      return Checksum % 10 == 0;
    }
  }

  public static long Create(long number)
  {
    number *= 10;
    Luhn luhn = new Luhn(number);
    return luhn.Valid ? number : number + (10 - (luhn.Checksum % 10));
  }

  private int[] Doubled(int[] input)
  {
    int[] output = input.ToArray();
    int i = output.Length % 2 == 0 ? 0 : 1;
    for (; i < output.Length - 1; i += 2)
    {
      int doubled = output[i] * 2;
      output[i] = doubled > 9 ? doubled - 9 : doubled;
    }
    return output;
  }
}
